openproblems/tasks/spatial_decomposition/methods/cell2location.py

```diff
@@ -0,0 +1,249 @@
+            [
+                f"means_per_cluster_mu_fg_{i}"
+                for i in adata_sc.uns["mod"]["factor_names"]
+            ]
+        ].copy()
+    else:
+        inf_aver = adata_sc.var[
+            [
+                f"means_per_cluster_mu_fg_{i}"
+                for i in adata_sc.uns["mod"]["factor_names"]
+            ]
+        ].copy()
+    inf_aver.columns = adata_sc.uns["mod"]["factor_names"]
+
+    # SPATIAL MAPPING
+    # find shared genes and subset both anndata and reference signatures
+    intersect = np.intersect1d(adata.var_names, inf_aver.index)
+    adata = adata[:, intersect].copy()
+    inf_aver = inf_aver.loc[intersect, :].copy()
+
+    # prepare anndata for cell2location model
+    adata.obs["sample"] = "all"
+    Cell2location.setup_anndata(adata=adata, batch_key="sample")
+    if not amortised:
+        # create and train the model
+        st_model = Cell2location(
+            adata,
+            cell_state_df=inf_aver,
+            # the expected average cell abundance: tissue-dependent
+            # hyper-prior which can be estimated from paired histology:
+            # here = average in the simulated dataset
+            N_cells_per_location=N_cells_per_location,
+            # hyperparameter controlling normalisation of
+            # within-experiment variation in RNA detection:
+            detection_alpha=detection_alpha,
+        )
+
+        st_model.train(
+            max_epochs=max_epochs_sp,
+            # train using full data (batch_size=None)
+            batch_size=st_batch_size,
+            # use all data points in training because
+            # we need to estimate cell abundance at all locations
+            train_size=1,
+        )
+    else:
+        # create and train the model
+        st_model = Cell2location(
+            adata,
+            cell_state_df=inf_aver,
+            # the expected average cell abundance: tissue-dependent
+            # hyper-prior which can be estimated from paired histology:
+            # here = average in the simulated dataset
+            N_cells_per_location=N_cells_per_location,
+            # hyperparameter controlling normalisation of
+            # within-experiment variation in RNA detection:
+            detection_alpha=detection_alpha,
+            amortised=True,
+            encoder_mode="multiple",
+            encoder_kwargs={
+                "dropout_rate": 0.1,
+                "n_hidden": {
+                    "single": 256,
+                    "n_s_cells_per_location": 10,
+                    "b_s_groups_per_location": 10,
+                    "z_sr_groups_factors": 64,
+                    "w_sf": 256,
+                    "detection_y_s": 20,
+                },
+                "use_batch_norm": False,
+                "use_layer_norm": True,
+                "n_layers": 1,
+                "activation_fn": ELU,
+            },
+        )
+
+        st_model.train(
```

Comment:
Looks like this code is duplicated between amortised=True and False. Can we move it outside the if statement?